scenes/test_array.txt

```diff
@@ -241,6 +241,239 @@
+*set test_arr_1 "1"
+*set test_arr_2 "2"
+*set test_arr_3 "3"
+*set test_arr_4 "4"
+*set test_arr_count 4
+*gosub_scene cslib_array test_all "test_arr" "test_array" "test_all_is_positive"
+*gosub test_finish true cslib_ret
+
+*gosub test_start "test_all #3 (empty)"
+*set test_arr_count 0
+*gosub_scene cslib_array test_all "test_arr" "test_array" "test_all_is_positive"
+*gosub test_finish true cslib_ret
+
+*gosub test_start "test_all #4 (complex, fail)"
+*set test_arr_1 "A"
+*set test_arr_2 "A"
+*set test_arr_3 "B"
+*set test_arr_4 "B"
+*set test_arr_5 "C"
+*set test_arr_6 "D"
+*set test_arr_count 6
+*gosub_scene cslib_array test_all "test_arr" "test_array" "test_all_duplicate_line"
+*if (false)
+	*comment check every 2nd element is identical
+	*comment to the one before it
+	*label test_all_duplicate_line
+	*params value index arr_ref
+	*set cslib_ret false
+	*comment ignore odd lines
+	*if not((index modulo 2) = 0)
+		*set cslib_ret true
+	*elseif ({arr_ref&"_${index}"} = {arr_ref&"_${index - 1}"})
+		*set cslib_ret true
+	*return
+*gosub test_finish false cslib_ret
+
+*gosub test_start "test_all #4 (complex, pass)"
+*set test_arr_1 "A"
+*set test_arr_2 "A"
+*set test_arr_3 "B"
+*set test_arr_4 "B"
+*set test_arr_5 "C"
+*set test_arr_6 "C"
+*set test_arr_count 6
+*gosub_scene cslib_array test_all "test_arr" "test_array" "test_all_duplicate_line"
+*gosub test_finish true cslib_ret
+
+*comment :::::: TEST_SOME tests ::::::
+*gosub test_start "test_some #1 (fail)"
+*set test_arr_1 "lost"
+*set test_arr_2 "lost"
+*set test_arr_3 "lost"
+*set test_arr_4 "lost"
+*set test_arr_count 4
+*gosub_scene cslib_array test_some "test_arr" "test_array" "test_some_won_match"
+*if (false)
+	*label test_some_won_match
+	*params match_result index
+	*set cslib_ret false
+	*if (match_result = "won")
+		*set cslib_ret true
+	*return
+*gosub test_finish false cslib_ret
+
+*gosub test_start "test_some #2 (success)"
+*set test_arr_1 "lost"
+*set test_arr_2 "won"
+*set test_arr_3 "lost"
+*set test_arr_4 "won"
+*set test_arr_count 4
+*gosub_scene cslib_array test_some "test_arr" "test_array" "test_some_won_match"
+*gosub test_assert_equal true cslib_ret
+*comment we're still a winner if we shrink the array
+*set test_arr_count 3
+*gosub_scene cslib_array test_some "test_arr" "test_array" "test_some_won_match"
+*gosub test_assert_equal true cslib_ret
+*gosub test_finish
+
+*gosub test_start "test_some #3 (empty)"
+*set test_arr_count 0
+*gosub_scene cslib_array test_some "test_arr" "test_array" "test_some_won_match"
+*gosub test_finish false cslib_ret
+
+*gosub test_start "test_some #4 (complex, fail)"
+*set test_arr_1 "museum"
+*set test_arr_2 "palindrome"
+*set test_arr_3 "loop"
+*set test_arr_4 "software"
+*set test_arr_5 "holiday"
+*comment civic, radar, level, rotor, kayak
+*set test_arr_count 5
+*gosub_scene cslib_array test_some "test_arr" "test_array" "is_palindrome"
+*if (false)
+	*comment check if a word spells itself backwards
+	*label is_palindrome
```

Comment:
do you want to move this to the string module? It's a cute example and people may use it for a magic incantation.
*text_input password
*if is_palindrome...  (not real CS)

Reply:
So for the "complex" examples, I very nearly did call some cslib_string routines, but then I remembered what you said about not having tests rely on the functionality of other modules (which was a good point).

The palindrome routine is fun, but I don't know if it's useful enough to warrant a place in cslib_string itself.
It's kind of a slippery slope to having all sorts of not-very useful (day-to-day) stuff to maintain and run tests.

Reply:
good point